### ISSUE TITLE
Allow duplicated package names

### ISSUE THREAD

**Issue by mhsmith** [beeware/briefcase#975]
If any of the `sources` lists contain directories with the same basenames, Briefcase gives an error like this:
```
The `sources` list for 'toga-test' contains duplicated package names.
```

However, there are legitimate reasons for wanting to merge together multiple directories into a single package. There's even a [PEP about it](https://peps.python.org/pep-0420/). And it would make the Toga test code organization a bit simpler if both the common code and the backend-specific code were in the same `tests` package.

The [Briefcase documentation](https://briefcase.readthedocs.io/en/latest/reference/configuration.html#sources) already indicates how this would work:
> If an application defines sources at the global level, application level, and platform level, the final set of sources will be the concatenation of sources from all levels, starting from least to most specific.

In other words:
* For `briefcase dev`, the directories would be added to the PYTHONPATH in that order. This should already work: we just need to remove the prohibition on duplicate directory names.
* For `briefcase run`, the directories would be copied into the app in that order, with later files overwriting earlier ones, just as pip does. Implementing this might be as simple as passing the `dirs_exist_ok=True` option to `shutil.copytree`.

**Comment by mhsmith** [comment]
I've changed my mind about using this in the test app: see https://github.com/beeware/toga/pull/1687#discussion_r1035482562. So there's no immediate need to fix this. But when we do, it's important to minimize any differences between `briefcase dev` and `briefcase run`. Here's how we can do that:

* For `briefcase dev`, the directories would be added to the PYTHONPATH in order from most to least specific, so most specific takes priority.
* For `briefcase run`, the directories would be copied into the app in order from most to least specific:
  * Files with the same path would *not* overwrite, and would give a warning if they had different content.
  * Directories with the same path would be merged. The only cases which guarantee consistency with `briefcase dev` are when either none of the directories contain an `__init__.py` file (an implicit namespace package), or the first directory contains an `__init__.py` file with a call to [`pkgutil.extend_path`](https://docs.python.org/3/library/pkgutil.html#pkgutil.extend_path), and the other directories' `__init__.py` files are either identical or missing. All other cases should give a warning.

**Comment by freakboy3742** [comment]
An implementation note - if/when this is done, there are some paths that *shouldn't* be copied as part of a merge process - see #986 for a discussion.
